[shard] make state_dict hook be consistent

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79650

for root module we shouldn't accidentally add a "." for state_dict keys, it
should be empty instead to match the module.state_dict behavior

Differential Revision: [D37191203](https://our.internmc.facebook.com/intern/diff/D37191203)

wanchaol committed:
[shard] make state_dict hook be consistent

for root module we shouldn't accidentally add a "." for state_dict keys, it
should be empty instead to match the module.state_dict behavior

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37191203).

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1097,7 +1097,11 @@ def test_state_dict(self):
         m._register_state_dict_hook(state_dict_hook)
         buffer = io.BytesIO()
-        torch.save(m.state_dict(), buffer)
+        mod_state_dict = state_dict = m.state_dict()
+        mod_state_keys = mod_state_dict.keys()
+        self.assertTrue("sharded_tensor1" in mod_state_keys)
+        self.assertTrue("submodule.sharded_tensor2" in mod_state_keys)
+        torch.save(mod_state_dict, buffer)
```
Would this cause the unit test to fail without the changes below? Also, shouldn't we add test coverage for this to the load path as well since we made some changes there?

fduwjj commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1097,7 +1097,11 @@ def test_state_dict(self):
         m._register_state_dict_hook(state_dict_hook)
         buffer = io.BytesIO()
-        torch.save(m.state_dict(), buffer)
+        mod_state_dict = state_dict = m.state_dict()
```
Nit: maybe just mod_state_dict = m.state_dict()?

fduwjj reviewed (approved):
lgtm

wanchaol committed:
Update on "[shard] make state_dict hook be consistent"


for root module we shouldn't accidentally add a "." for state_dict keys, it
should be empty instead to match the module.state_dict behavior

Differential Revision: [D37191203](https://our.internmc.facebook.com/intern/diff/D37191203)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37191203).

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37191203).

wanchaol committed:
Update on "[shard] make state_dict hook be consistent"


for root module we shouldn't accidentally add a "." for state_dict keys, it
should be empty instead to match the module.state_dict behavior

Differential Revision: [D37191203](https://our.internmc.facebook.com/intern/diff/D37191203)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37191203).

wanchaol committed:
Update on "[shard] make state_dict hook be consistent"


for root module we shouldn't accidentally add a "." for state_dict keys, it
should be empty instead to match the module.state_dict behavior

Differential Revision: [D37191203](https://our.internmc.facebook.com/intern/diff/D37191203)

[ghstack-poisoned]

wanchaol committed:
Update on "[shard] make state_dict hook be consistent"


for root module we shouldn't accidentally add a "." for state_dict keys, it
should be empty instead to match the module.state_dict behavior

Differential Revision: [D37191203](https://our.internmc.facebook.com/intern/diff/D37191203)

[ghstack-poisoned]
